SimulationAgent/Simulation/DeviceStatusLogic/UpdateReportedProperties.cs

```diff
@@ -4,41 +4,97 @@
 using System.Collections.Generic;
 using System.Threading;
 using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Concurrency;
 using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Diagnostics;
 using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Models;
 using Microsoft.Azure.IoTSolutions.DeviceSimulation.SimulationAgent.Exceptions;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.SimulationAgent.Simulation.DeviceStatusLogic.Models;
 
 namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.SimulationAgent.Simulation.DeviceStatusLogic
 {
     public class UpdateReportedProperties : IDeviceStatusLogic
     {
-        // When connecting to IoT Hub, timeout after 10 seconds
-        private static readonly TimeSpan connectionTimeout = TimeSpan.FromSeconds(10);
+        // Twin write timeout
+        private static readonly TimeSpan updateTwinTimeout = TimeSpan.FromSeconds(10);
 
-        private IDevices devices;
+        // Twin update frequency
+        private static readonly TimeSpan reportedPropertyUpdateFrequency = TimeSpan.FromSeconds(30);
+
+        private readonly IDevices devices;
         private string deviceId;
         private DeviceModel deviceModel;
 
+        // The timer invoking the Run method
+        private readonly ITimer timer;
+
         private readonly ILogger log;
 
         // Ensure that setup is called once and only once (which helps also detecting thread safety issues)
         private bool setupDone = false;
 
+        private IDeviceActor context;
+
         public UpdateReportedProperties(
+            ITimer timer,
             IDevices devices,
             ILogger logger)
         {
+            this.timer = timer;
             this.log = logger;
             this.devices = devices;
         }
 
+        public void Setup(string deviceId, DeviceModel deviceModel, IDeviceActor context)
+        {
+            if (this.setupDone)
+            {
+                this.log.Error("Setup has already been invoked, are you sharing this instance with multiple devices?",
+                    () => new { this.deviceId });
+                throw new DeviceActorAlreadyInitializedException();
+            }
+
+            this.setupDone = true;
+            this.deviceId = deviceId;
+            this.deviceModel = deviceModel;
+
+            this.context = context;
+            this.timer.Setup(this.Run, reportedPropertyUpdateFrequency);
+        }
+
+        public void Start()
+        {
+            this.log.Info("Starting UpdateReportedProperties timer",
+                () => new { this.context.DeviceId });
+            this.timer.Start();
+        }
+
+        public void Stop()
+        {
+            this.log.Info("Stopping UpdateReportedProperties timer",
```

Comment:
do you want to move these log entries into the timer.start, stop, pause, resume methods?  It'd save you copying them to each method.

Reply:
maybe in the future, I like them here for the time being.